### PR TITLE
feat: add pre-mine spend

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
Added pre-mine spend with improved logging for fault finding

Motivation and Context
---
Pre-mine spending did not work.

How Has This Been Tested?
---
A new esmeralda genesis block that includes pre-mine with a 2-of-3 scheme and a backup spending wallet was created, utilizing 4x ledger console wallets. Those wallets were then used to spend the 1st pre-mine UTXO. The UTXO was successfully mined and imported into the target wallet.

Leader's ledger console wallet console:
```
PS C:\Users\pluto\.tari-Hansie\node_01> .\minotari_console_wallet.exe --base-path . --network esmeralda --config $pwd"\config\config.toml" --log-config $pwd"\config\log4rs_console_wallet.yml" pre-mine-spend-session-info --fee-per-gram 5 --output-index 0 --recipient-address f27jozcSQjY2QfopvAeuNMDJX5z9T6vThdBRSteaanwecG1F31ez7vUUCVCihFL6UrJHLVY9EPxkRrgHJLwdPJNxtz4 --verify-unspent-outputs
Initializing logging according to "C:\\Users\\pluto\\.tari-Hansie\\node_01\\config\\log4rs_console_wallet.yml"
Minotari Console Wallet running... (Command mode started)
==============
Command Runner
==============

1. PreMineSpendSessionInfo(PreMineSpendSessionInfoArgs { fee_per_gram: MicroMinotari(5), output_index: 0, recipient_address: Dual(DualAddress { network: Esmeralda, features: TariAddressFeatures(1), public_view_key: 4c5b2793fc2504965138bebf97862af0865ff249e3b8e69b06d441dab0f70346, public_spend_key: b492d5ddf85688c452765180b6e18cc31a0a3ee401ec275d67118145d97cbd67 }), verify_unspent_outputs: true })


Concluded step 1 'pre-mine-generate-session-info'
Your session ID is:                 '8Tj5eMbr7Z3NXUs3'
Your session's output directory is: 'C:\Users\pluto\Documents\tari_pre_mine\spend\8Tj5eMbr7Z3NXUs3'
Session info saved to:              'C:\Users\pluto\Documents\tari_pre_mine\spend\8Tj5eMbr7Z3NXUs3\step_1_session_info.json'
Send 'step_1_session_info.json' to parties for step 2

Minotari Console Wallet running... (Command mode completed)

Shutting down wallet... Done.
PS C:\Users\pluto\.tari-Hansie\node_01> .\minotari_console_wallet.exe --base-path . --network esmeralda --config $pwd"\config\config.toml" --log-config $pwd"\config\log4rs_console_wallet.yml" pre-mine-spend-encumber-aggregate-utxo --session-id 8Tj5eMbr7Z3NXUs3 --input-file-names=step_2_for_leader_from_carol.json
Initializing logging according to "C:\\Users\\pluto\\.tari-Hansie\\node_01\\config\\log4rs_console_wallet.yml"
Minotari Console Wallet running... (Command mode started)
==============
Command Runner
==============

1. PreMineSpendEncumberAggregateUtxo(PreMineSpendEncumberAggregateUtxoArgs { session_id: "8Tj5eMbr7Z3NXUs3", input_file_names: ["step_2_for_leader_from_carol.json"] })


Concluded step 3 'pre-mine-encumber-aggregate-utxo'
Send 'step_3_for_parties.json' to parties for step 4

Minotari Console Wallet running... (Command mode completed)

Shutting down wallet... Done.
PS C:\Users\pluto\.tari-Hansie\node_01> .\minotari_console_wallet.exe --base-path . --network esmeralda --config $pwd"\config\config.toml" --log-config $pwd"\config\log4rs_console_wallet.yml" pre-mine-spend-aggregate-transaction --session-id 8Tj5eMbr7Z3NXUs3 --input-file-names=step_4_for_leader_from_carol.json
Initializing logging according to "C:\\Users\\pluto\\.tari-Hansie\\node_01\\config\\log4rs_console_wallet.yml"
Minotari Console Wallet running... (Command mode started)
==============
Command Runner
==============

1. PreMineSpendAggregateTransaction(PreMineSpendAggregateTransactionArgs { session_id: "8Tj5eMbr7Z3NXUs3", input_file_names: ["step_4_for_leader_from_carol.json"] })


Concluded step 5 'pre-mine-spend-aggregate-utxo'

Minotari Console Wallet running... (Command mode completed)

Shutting down wallet... Done.
PS C:\Users\pluto\.tari-Hansie\node_01>
```
Party member's ledger console wallet console:
```
PS C:\Users\pluto\.tari-Hansie\node_03> .\minotari_console_wallet.exe --base-path . --network esmeralda --config $pwd"\config\config.toml" --log-config $pwd"\config\log4rs_console_wallet.yml" pre-mine-spend-party-details --alias carol --output-index 0 --input-file "C:\Users\pluto\Documents\tari_pre_mine\spend\8Tj5eMbr7Z3NXUs3\step_1_session_info.json"
Initializing logging according to "C:\\Users\\pluto\\.tari-Hansie\\node_03\\config\\log4rs_console_wallet.yml"
Minotari Console Wallet running... (Command mode started)
==============
Command Runner
==============

1. PreMineSpendPartyDetails(PreMineSpendPartyDetailsArgs { input_file: "C:\\Users\\pluto\\Documents\\tari_pre_mine\\spend\\8Tj5eMbr7Z3NXUs3\\step_1_session_info.json", output_index: 0, alias: "carol" })


Concluded step 2 'pre-mine-create-party-details'
Your session's output directory is 'C:\Users\pluto\Documents\tari_pre_mine\spend\8Tj5eMbr7Z3NXUs3'
Send 'step_2_for_leader_from_carol.json' to leader for step 3

Minotari Console Wallet running... (Command mode completed)

Shutting down wallet... Done.
PS C:\Users\pluto\.tari-Hansie\node_03> .\minotari_console_wallet.exe --base-path . --network esmeralda --config $pwd"\config\config.toml" --log-config $pwd"\config\log4rs_console_wallet.yml" pre-mine-spend-input-output-sigs --session-id 8Tj5eMbr7Z3NXUs3
Initializing logging according to "C:\\Users\\pluto\\.tari-Hansie\\node_03\\config\\log4rs_console_wallet.yml"
Minotari Console Wallet running... (Command mode started)
==============
Command Runner
==============

1. PreMineSpendInputOutputSigs(PreMineSpendInputOutputSigArgs { session_id: "8Tj5eMbr7Z3NXUs3" })


Concluded step 4 'pre-mine-create-input-output-sigs'
Send 'step_4_for_leader_from_carol.json' to leader for step 5

Minotari Console Wallet running... (Command mode completed)

Shutting down wallet... Done.
PS C:\Users\pluto\.tari-Hansie\node_03>
```

What process can a PR reviewer use to test or verify this change?
---
Code review
System-level testing

Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

File: applications/minotari_ledger_wallet/wallet/src/handlers/get_script_offset.rs [hunk 1]
```diff
@@ -57,89 +63,139 @@ impl ScriptOffsetCtx {
...
+        offset_ctx.script_private_key_sum =
+            Zeroizing::new(offset_ctx.script_private_key_sum.deref() + script_key.deref());
+    }
+
+    // 5. Derived sender offsets key
+    let end_derived_offset_keys = end_script_indexes + offset_ctx.total_derived_offset_keys;
+    if (end_script_indexes..end_derived_offset_keys).contains(&(chunk_number as u64)) {
+        let k = derive_key_from_alpha(offset_ctx.account, data)?;
+
+        offset_ctx.add_unique_key(k.clone());
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/wallet/src/handlers/get_script_offset.rs]: We should not count K as a unique key here, 
The real unique key here is alpha aka the spend key. 
K is calculated as `k = spend_pvt + Hash(key)`
Having 2 unique `k` keys you can still calculate the `spend_pvt` key 

Solution would be to add spend here.

File: applications/minotari_ledger_wallet/wallet/src/handlers/get_script_offset.rs [hunk 2]
```diff
@@ -57,89 +63,139 @@ impl ScriptOffsetCtx {
...
+        offset_ctx.sender_offset_sum = Zeroizing::new(offset_ctx.sender_offset_sum.deref() + k.deref());
+    }
 
-        let k = alpha_hasher(alpha, blinding_factor)?;
+    // 6. Derived script key
+    let end_derived_script_keys = end_derived_offset_keys + offset_ctx.total_derived_script_keys;
+    if (end_derived_offset_keys..end_derived_script_keys).contains(&(chunk_number as u64)) {
+        let k = derive_key_from_alpha(offset_ctx.account, data)?;
 
         offset_ctx.add_unique_key(k.clone());
```
Comment (SWvheerden) [inline applications/minotari_ledger_wallet/wallet/src/handlers/get_script_offset.rs]: We should not count K as a unique key here, 
The real unique key here is alpha aka the spend key. 
K is calculated as `k = spend_pvt + Hash(key)`
Having 2 unique `k` keys you can still calculate the `spend_pvt` key 

Solution would be to add spend here.

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 3]
```diff
@@ -787,31 +822,57 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                        TariKeyId::Managed { branch, index } => {
+                            let km = self
+                                .key_managers
+                                .get(branch)
+                                .ok_or_else(|| self.unknown_key_branch_error("get_script_signature", branch))?
+                                .read()
+                                .await;
+                            km.get_private_key(*index)
+                                .map_err(|e| TransactionError::KeyManagerError(e.to_string()))?
+                        },
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]: this is incorrect
If the key type is managed, we need to send over an index, not the pvt key, or we need to sign locally. 
If we send over the managed key, the ledger is going to treat it as a derived key, which is not correct.

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 4]
```diff
@@ -958,31 +1015,42 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+        //             WalletType::Ledger(_) => {
+        //                 let key = TariKeyId::from_str(key.to_string().as_str())
+        //                     .map_err(|_| KeyManagerServiceError::KeySerializationError)?;
+        //
+        //                 let key = self.get_private_key(&key).await?;
+        //                 derived_key_commitments.push(key);
+        //             },
+        //         },
+        //     }
+        // }
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]: ?

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 5]
```diff
@@ -1003,26 +1070,92 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
                 {
-                    let mut total_script_private_key = PrivateKey::default();
+                    let mut partial_script_offset = PrivateKey::default();
+                    let mut derived_script_keys = vec![];
+                    let mut script_key_indexes = vec![];
+                    for script_key_id in script_key_ids {
+                        match script_key_id {
+                            TariKeyId::Managed { branch, index } => {
+                                match TransactionKeyManagerBranch::from_key(branch) {
+                                    TransactionKeyManagerBranch::Spend | TransactionKeyManagerBranch::PreMine => {
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
                                    TransactionKeyManagerBranch::Spend | TransactionKeyManagerBranch::PreMine| TransactionKeyManagerBranch::Random| TransactionKeyManagerBranch::SenderOffsetOneSided => {
```

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 6]
```diff
@@ -1003,26 +1070,92 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                                        script_key_indexes
+                                            .push((TransactionKeyManagerBranch::from_key(branch), *index));
+                                    },
+                                    _ => {
+                                        return Err(TransactionError::from(
+                                            self.branch_not_supported_error("get_script_offset", branch),
+                                        ));
+                                    },
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
                                    _ => {
                                        partial_script_offset += self.get_private_key(script_key_id).await? 
                                    },
```
This is incorrect. Keymanager should not force key types used. This should be up to the higher level logic

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 7]
```diff
@@ -1003,26 +1070,92 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                                }
+                            },
+                            TariKeyId::Derived { key } => {
+                                // TODO: What about if the derived key is a TariKeyId::Managed, but
+                                // TODO: `self.get_private_key(...)` will error?
+                                let key_id = TariKeyId::from_str(key.to_string().as_str())
+                                    .map_err(|_| KeyManagerServiceError::KeySerializationError)?;
+                                let k = self.get_private_key(&key_id).await?;
+                                derived_script_keys.push(k);
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
                                let key_id = TariKeyId::from_str(key.to_string().as_str())
                                    .map_err(|_| KeyManagerServiceError::KeySerializationError)?;
                                let k = self.get_private_key(&key_id).await?;
                                derived_script_keys.push(k);
```
if self.get_private_key(...) errors, then the higher logic assigned the wrong key. Its fine if we return the error form get_private_key

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 8]
```diff
@@ -1003,26 +1070,92 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                            TariKeyId::Imported { .. } => {
+                                partial_script_offset =
+                                    &partial_script_offset + self.get_private_key(script_key_id).await?
+                            },
+                            TariKeyId::Zero => {
+                                return Err(self.key_id_not_supported_error(
+                                    "get_script_offset",
+                                    "KeyId::Managed or TariKeyId::Imported",
+                                    script_key_id,
+                                ));
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
     {}
```
Just do nothing, it should not be an error. 
Adding 0 here should be fine

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 9]
```diff
@@ -1003,26 +1070,92 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                                    TransactionKeyManagerBranch::OneSidedSenderOffset |
+                                    TransactionKeyManagerBranch::RandomKey => {
+                                        sender_offset_indexes
+                                            .push((TransactionKeyManagerBranch::from_key(branch), *index));
+                                    },
+                                    _ => {
+                                        return Err(TransactionError::from(
+                                            self.branch_not_supported_error("get_script_offset", branch),
+                                        ));
+                                    },
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]: Same thing as above, dont impose limitations on the keys here

Comment (SWvheerden) [reply base_layer/core/src/transactions/key_manager/inner.rs]: It should support all valid keys

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 10]
```diff
@@ -1003,26 +1070,92 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
                                 }
                             },
-                            TariKeyId::Derived { .. } | TariKeyId::Imported { .. } | TariKeyId::Zero => {},
+                            TariKeyId::Derived { key } => {
+                                // TODO: What about if the derived key is a TariKeyId::Managed, but
+                                // TODO: `self.get_private_key(...)` will error?
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
```

File: base_layer/core/src/transactions/key_manager/inner.rs [hunk 11]
```diff
@@ -1003,26 +1070,92 @@ where TBackend: KeyManagerBackend<PublicKey> + 'static
...
+                            TariKeyId::Imported { .. } => {
+                                partial_script_offset =
+                                    partial_script_offset - self.get_private_key(sender_offset_key_id).await?;
+                            },
+                            TariKeyId::Zero => {
+                                return Err(self.key_id_not_supported_error(
+                                    "get_script_offset",
+                                    "KeyId::Managed or TariKeyId::Imported",
+                                    sender_offset_key_id,
+                                ));
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/key_manager/inner.rs]:
```suggestion
{}
```

**Review by SWvheerden — CHANGES_REQUESTED** [review]
Key manager should not impose limitations on which keys it must use where. 
That must be decided by the calling logic. 
The key manager should look at the key type and decide what to do.

**Review by SWvheerden — DISMISSED** [review]

**Comment by brianp** [comment]
I haven't reviewed the changes yet but it is functional:

All scenarios produce spendable outputs.
✅ One-sided: Ledger -> Ledger
✅ One-sided: Ledger -> Software
✅ One-sided: Coinbase -> Ledger
✅ One-sided: Software -> Ledger
✅ Interactive: Software -> Ledger

**Pull request merged**